src/mesh/replicated_mesh.C

```diff
@@ -173,6 +173,38 @@ ReplicatedMesh::ReplicatedMesh (const UnstructuredMesh & other_mesh) :
       other_boundary_info.get_nodeset_name(node_bnd_id);
 }
 
+ReplicatedMesh & ReplicatedMesh::operator= (ReplicatedMesh && other_mesh)
+{
+  // Move assign as an UnstructuredMesh
+  this->UnstructuredMesh::operator=(std::move(other_mesh));
+
+  // Nodes and elements belong to ReplicatedMesh and have to be
+  // moved before we can move arbitrary GhostingFunctor, Partitioner,
+  // etc. subclasses.
+  this->move_nodes_and_elements(std::move(other_mesh));
+
+  // Handle those remaining moves.
+  this->post_dofobject_moves(std::move(other_mesh));
+
+  return *this;
+}
+
+MeshBase & ReplicatedMesh::assign(MeshBase && other_mesh)
+{
+  *this = std::move(*(cast_ptr<ReplicatedMesh*>(&other_mesh)));
+
+  return *this;
+}
+
+void ReplicatedMesh::move_nodes_and_elements(MeshBase && other_mesh)
+{
+  this->_nodes = std::move((cast_ptr<ReplicatedMesh*>(&other_mesh))->_nodes);
```

Comment:
Rather than repeated checked casts, just do one at the beginning of the function and use that reference.